### ISSUE TITLE
Change fields (in Json profile, display screen and Bag-info) from set to ordered list

### ISSUE THREAD

**Issue by johnscancella** [LibraryOfCongress/bagger#19]
The current set does not allow desired ordering of metadata entry fields (in Json profile, display screen and Bag-info). It is is available here: [https://github.com/LibraryOfCongress/bagger/blob/master/bagger/src/main/java/gov/loc/repository/bagger/ui/BagInfoForm.java#L86]

Desired improvement:
For an ordered field of metadata to be possible for accessions of digital records that will hep make sense of the keyed information in Bag-Info and on the display screen, it is desired that an ordered list be implemented. With such an improvement, profiles will help generate understandable metadata in the order originally intended for during project profile creation.
This improvement will also help make Bagger very usable and customizable. Additionally, importing SIPs created into preservation or records systems will be facilitated by the metadata structure/order on mapping fields between Bagger (Bag-info) and the new system's metadata structure.

**Comment by johnscancella** [comment]
note: this was created for @houzanme1.

**Comment by johnscancella** [comment]
Looking at the bagit spec for [metadata](https://tools.ietf.org/html/draft-kunze-bagit-13#section-2.2.2) it does specifically state that the order should be preserved which in this case bagger does not currently comply with

**Comment by acdha** [comment]
@johnscancella: can this be as simple as changing https://github.com/LibraryOfCongress/bagger/blob/master/bagger/src/main/java/gov/loc/repository/bagger/ui/BagInfoForm.java#L44 to use a LinkedHashMap to display the fields in insertion order?

**Comment by johnscancella** [comment]
For the ordering yes, however the spec also says there may be duplicates (through in practical usage that may never happ). I was just about to test LinkedHashMap (great minds think alike a guess)

**Comment by johnscancella** [comment]
@houzanme1 the [push](https://github.com/LibraryOfCongress/bagger/commit/e1e5b7f0f9a7e122b7d13a2641b0cce38e207fe0) I just made should solve the ordering issue for you, but still isn't to the bagit spec since it doesn't allow multiples of the same field. 

If you are feeling adventurous you can try building and using the latest code until the next release is pushed out. Just follow the [build process from the README](https://github.com/LibraryOfCongress/bagger#bagger-build-process)

**Comment by houzanme1** [comment]
@johnscancella and @acdha : Well guys, that was faster than I hoped for!
Thanks to you both for acting on this. Will try and let you know if I succeed in making it work on my end.

Can't wait for the next release. BTW, will send a cleaner profile before that release if you tell me approximately when it is due out.
Thanks!
Tibaut

**Comment by johnscancella** [comment]
@houzanme1 @dbrunton or @kzwa can give a better answer as to when the next release is

**Comment by houzanme1** [comment]
Does the fix assume I have GitHub for Windows to build and use the latest code with Gradle? (before the next release)

**Comment by johnscancella** [comment]
@houzanme1 I am not sure I understand. GitHub for Windows is just a git client and has nothing to do with the code. Git is just the source code repository format, any git compatible tool should be able to get the code from github. 

Once you have the code on your computer, then use gradle to build.

Or you can download [release 2.5 candidate 2](https://github.com/LibraryOfCongress/bagger/releases/tag/v2.5-RC2) which has already been built

**Comment by houzanme1** [comment]
I just meant to say, I am limited in what programs I can install and what I can do on a restricted computer. I am super glad  release 2.5 candidate 2  is available!

As for the original issue, the fields order in the Json profile still does not match Bagger's display of the same fields. I did notice a new set of fields that can be added from the "standard" list. 

The profile I am submitting below (prepared for our agency), is the synthesis of what appears to be important when describing and accessioning digital records. It should make accessions easier by eliminating the need to think about what field is important to add. Eliminating is easier...during accession time....:-)

Though many fields are suggested, only five (5) are required to properly identify the record and transfer/accession. 

This profile can be adapted/adopted as users wish, though I have not been able to add comment lines and contact information to follow up with either change requests or improvements (Bagger does not recognize the profile any more when //  or /*...*/ is added)...

Here is the profile named IARA(Indiana)-Accession-Profile-profile.json (See attached file saved in .txt --would need proper extension-- if code turns out to be mal-formed as it can be when moved around :-( )


[IARA(Indiana)-Accession-Profile-profile.txt](https://github.com/LibraryOfCongress/bagger/files/133315/IARA.Indiana.-Accession-Profile-profile.txt)


**Json Profile content copied/pasted here:**


{

	"Accession/Transfer Number": {
		"fieldRequired": true

	},
	"Transferring-Agency/Code": {
		"fieldRequired": true
	},
	"Subdivision/Code": {
		"fieldRequired": false
	},
	"Agency/Subdivision-Adress": {
		"fieldRequired": false
	},
	"Transferring-Employee": {
		"fieldRequired": false
	},

	"Receiving-Employee": {
		"fieldRequired": false
	},

	"RecordsSeriesNumber": {
		"fieldRequired": false

	},
	"RecordsSeriesTitle": {
		"fieldRequired": true
	},
	"RecordsSeriesRetention": {
		"fieldRequired": false,
		"valueList": ["Permanent(Keep-For-Ever)", "Non-Permanent(Timed-Retention)", "Not-yet-Known"]

	},

	"RecordsTitle/Description": {
		"fieldRequired": true


	},

	"Records-Medium": {
		"filedRequired": true,
		"defaultValue": "CD",
		"valueList": ["CD", "DVD", "Blu-Ray", "Hard-Drive", "Usb-Flash/Thumb-Drive",
			"Computer",
			"Folder-on-Computer",
			"Email/Attachment",
			"API/CMIS/APP",
			"Content/Records-Systems",
			"FTP/SFTP/Network-Location",
			"Floppy - 3.5 in ",
			"Floppy - 5.25 in ",
			"Floppy - 8 in ",
			"Other - describe - in notes "
		]

	},

	"Number-Of-Media-Shipped": {
		"fieldRequired": false
	},

	"Ship-Date": {
		"fieldRequired": true
	},
	"Ship-Tracking-Number": {
		"fieldRequired": false
	},
	"Ship-Method": {
		"fieldRequired": false
	},
	"Transfer-To-State-Archives": {
		"fieldRequired": false,
		"defaultValue": "Indiana Archives and Records Administration - 6444, E 30th Street, Indianapolis, IN 46219 "
	},


	"Birth-Means": {
		"filedRequired": true,
		"defaultValue": "Born-Digital",
		"valueList": ["Born-Digital", "Reformatted/Recaptured Digital", "Not-Yet-Known"]
	},


	"Originality/Uniqueness": {
		"filedRequired": false,
		"defaultValue": "???",
		"valueList": ["Original-By-Birth", "Surrogate(Physical/Original-Kept)", "Original-By-Default(Physical/OriginalDestroyed)",
			"No-Yet-Known"
		]
	},

	"Size-Before-Transfer": {
		"fieldRequired": false
	},

	"Hash/Signature-Before-Transfer": {
		"fieldRequired": false
	},


	"Hash/Signature-Type": {
		"fieldRequired": false,
		"defaultValue": "MD5",
		"valueList": ["MD5", "SHA1", "SHA224", "SHA256", "SHA-512", "SHA-384", "SHA-512", "SHA3-224", "SHA3-256", "SHA3-384", "SHA3-512"]

	},

	"Surrogate/Duplicate-Location": {
		"fieldRequired": false
	},


	"Classification-Legal/series": {
		"fieldRequired": false,
		"defaultValue": "???",
		"valueList": ["Open/Public", "Confidential/Sensitive", "Semi-Confidential", "Top-Secret/Classified", "Not-Yet-Known"]

	},

	"Classification-Judgement/preservation": {
		"fieldRequired": false,
		"defaultValue": "???",
		"valueList": ["Vital(Birth/Death/Marriage/Land/Rights)", "Critical(Indispensible for Business)",
			"High Value(Re-Appraised/Perceived)", "Not-Yet-Known"
		]

	},



	"Content-Type": {
		"filedRequired": true,
		"defaultValue": "mixed formats",
		"valueList": ["Word Processing", "Plain Text", "Text With MArkup", "Spreadsheet", "Presentation",

			"Database", "Audio-Mono", "Audio-Stereo", "Video-High-Quality-Professional", "Video-Medium-Quality-Amateur",
			"Video-Animation - or - Interactive ",
			"Image - Raster (for Tiff,PNG,Jpeg,jpeg2000,etc.)",
			"Image - Vector & CAD ",
			"Image - Raw/Native ",
			"Geospatial - Raster ",
			"Geospatial - Vector ",
			"Email",
			"WebPages/SocialMedia/Blogs",
			"Software/Game/Application/code",
			"Disk/Forensic Image",
			"Archive (Zip/Tar/Arc/Warc, etc.)",
			"Encrypted-With-key",
			"Encrypted-No-Key(content-Unknown)",
			"Mixed formats ",
			"No-Yet-Known"
		]
	},


	"Format Openness": {
		"fieldRequired": true,
		"valueList": ["OpenSource", "Proprietary", "Semi-Open/Proprietary", "Not-Yet-Known"]

	},


	"Preservation-Stage": {
		"fieldRequired": false,
		"defaultValue": "SIP",
		"valueList": ["SIP", "AIP", "DIP", "Encapsulated AIP"]
	},


	"Location-Content-Copied-To": {
		"fieldRequired": false
	},
	"Original-Media-Kept?": {
		"fieldRequired": false
	},
	"Original-Media-Location-if-Kept": {
		"fieldRequired": false
	},
	"Notes": {
		"fieldRequired": false
	}
}

**Comment by johnscancella** [comment]
after looking at the code some more it looks like the problem is actually upstream in [org.json.JSONObject line 157](https://github.com/stleary/JSON-java/blob/master/JSONObject.java#L157) where it is using a plain hashmap (does not preserve order). This JSONObject is used to convert the json text file into a java object. I will file a ticket with him and see if I can get him to update the code.

**Comment by johnscancella** [comment]
It looks like preserving order in the JSON object has been [discussed before](https://github.com/stleary/JSON-java/pull/190) and ultimately denied.

You may be able to get it to work if you change the JSON profile, but having not tested it I am not sure if it will work. See http://stackoverflow.com/a/4515863 for more details on how you could do this.

I think the real solution would be to rework/re-architect the profiles, but that would be a big change and would need more input from the general community that uses bagger.

**Comment by houzanme1** [comment]
If you get that to work, you would have solved the rood cause...and that
would be lovely!
Tibaut

Best Regards,

Tibaut Houzanme
-
Mobile: +1-317-332-3296
email: houzanme@gmail.com

   - www.linkedin.com/in/houzanme/

**Comment by johnscancella** [comment]
Yeah I was a little bummed about that, but I also understand that the spec states it is unordered so it is really an oversight by the original bagger creator. 

As for a heads up, if you have thoughts or problems feel free to post them here (after trying to work them out of course). If you change the json profiles to be an array named something, then you will probably have to update the `jsonBagger.java` code. Any pull requests are appreciated.

Otherwise I will keep this open so others know.

**Comment by houzanme1** [comment]
@johnscancella Sounds good. Will keep you posted.
Trying different profiles, sounds easier, and updating the java code might take a bit longer.
Thanks for leaving this open.

**Comment by houzanme1** [comment]
I tried various profiles suggested, without editing the jsonBagger.java
code. No success there.
Let me know if after what happens after you altered the code.
BTW, the drop-down list created by "valueList" in the profile is perfectly
ordered as entered. I am impressed with it.

Best Regards,

Tibaut Houzanme
-
Mobile: +1-317-332-3296
email: houzanme@gmail.com

   - www.linkedin.com/in/houzanme/

**Comment by acdha** [comment]
@johnscancella What do you thinking about using [`javax.json.stream`](https://docs.oracle.com/javaee/7/api/javax/json/stream/package-summary.html) and have it populate an ordered hash map as each event is processed?

**Comment by johnscancella** [comment]
@houzanme1 I implemented a compromise, if you change the profile to use an array which does enforce order. I changed yours to use an array and included it with the rest of the code. This will be in version 2.5-RC4

**Comment by johnscancella** [comment]
Ohh and in case you were still wondering about comments in the profile, by definition they are not allowed. See http://stackoverflow.com/a/4183018

**Comment by houzanme1** [comment]
That sounds great, @johnscancella !
My curiosity and excitement levels just went up. I can't wait ti see 2.5-RC-4!

Thanks for the heads up on comments. I thought explaining the rational for the fields would help get good feedback. No worries then. 

Let me know when RC4 comes out and I will be first to test it!

**Comment by johnscancella** [comment]
@houzanme1 test away https://github.com/LibraryOfCongress/bagger/releases/tag/v2.5-RC4

**Comment by houzanme1** [comment]
I tried the following short codes (arrays). None made the profiles appear for selection.
It would be nice if you could confirm if any of these work for you.
But, if you are able to create a very short one that works, share and I will model the larger profile on it.

Thanks
Tibaut

Test 1
---

{

"items":
	[
		{
			"w1":"que",
			"w2":"tre",
			"w3":"qie"
		},
	]

	}



Test2
---

{

"items":
	[
		
			{"Start": {"data": "que"}},
			{"Middle": {"data": "tre"}},
			{"End":" {"data": {"qie"}}
		
	]

}



Test3
---


{

"items":
	[
		
			{"Start": {"1": "que"}},
			{"Middle": {"2": "tre"}},
			{"End":" {"3": {"qie"}}
		
	]

	
}


Test 4
---

{

"items":
	[
		{
			"w1":"que",
			"w2":"tre",
			"w3":"qie"
		},
	],

"itemOrder":
	["w1","w2",w3"]

	}

**Comment by johnscancella** [comment]
You have to use the key `ordered`. There should be an example included with the new bagger called [ordered-other-project](https://github.com/LibraryOfCongress/bagger/blob/master/bagger-business/src/main/resources/gov/loc/repository/bagger/profiles/ordered-other-project-profile.json). I also included [your profile](https://github.com/LibraryOfCongress/bagger/blob/master/bagger-business/src/main/resources/gov/loc/repository/bagger/profiles/IARA-Indiana-Accession-profile.json) and edited it to already have the ordering.

**Comment by houzanme1** [comment]
I was able to load the new profile you ordered and it looks and works as pristine as ever! Very nice and satisfying work indeed!

As for the example you mentioned, I have not seen it when I launched RC4. Plus I am improving ours and requested comments from colleagues. Hopefully I get a real clean and beautified profile back for inclusion in future releases.

I can't say it well enough, but you have made an outstanding improvement to Bagger!

Tibaut
